Generic Inbound API Logging

esoergel opened the pull request:
## Product Description
no user-facing changes

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-2411
Adds log models from [the spec](https://docs.google.com/document/d/1lpGyVzur2G2lFOQJJ2nNiNQzasZFMccJiFtRxS5CSPo/edit#heading=h.aycnyl81jhmn), with the addition of `username` and `user_agent`.  It then hooks these models up to the main entry point for the API.
I left out `RequestLog.custom_metadata` for now, since it's marked "future".

I'd appreciate a close study of the model definitions - I made some judgement calls there that I'd like validated.

Once we implement retries, I'd like to rework the connection with API view - the view makes use of a bunch of bits of data from the request, and the log function attempts to capture those same bits, but it does it separately.  Since a retry will need access to exactly the same information as was present on the initial request, it'd be ideal in my mind to channel these two through a common interface to better enforce consistency.

## Feature Flag
Generic inbound APIs

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
I'm relying on automated tests here.  The feature is as-yet unreleased, so it should be pretty safe.

### Automated test coverage

included

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
N/A


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
This can be rolled back, but we'd have to revert the migration first.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Add generic inbound API logging models

esoergel committed:
Couple accepted request methods

The model choices is meant to reflect the methods allowed by the view,
so I figure it makes sense to link them together.  I'm open to
objections here, though.  Could also start with just POST and add others
as we need

esoergel committed:
Add __repr__s (feel free to quibble on the details)

esoergel committed:
Protect access to the generic API view

esoergel committed:
Log inbound API requests

esoergel committed:
We're also gonna need user agent for reprocessing

esoergel committed:
Add GIN index to case_ids field

>>> print(ProcessingAttempt.objects.filter(case_ids__contains=[uuid4()]).explain())
EXPLAIN SELECT "generic_inbound_processingattempt"."id",
       "generic_inbound_processingattempt"."log_id",
       "generic_inbound_processingattempt"."timestamp",
       "generic_inbound_processingattempt"."is_retry",
       "generic_inbound_processingattempt"."response_status",
       "generic_inbound_processingattempt"."response_body",
       "generic_inbound_processingattempt"."raw_response",
       "generic_inbound_processingattempt"."xform_id",
       "generic_inbound_processingattempt"."case_ids"
  FROM "generic_inbound_processingattempt"
 WHERE "generic_inbound_processingattempt"."case_ids" @>
ARRAY['14d34a63-b565-4925-a17c-416c901fdf9f'::uuid]::uuid[]

Execution time: 0.001054s [Database: default]
Bitmap Heap Scan on generic_inbound_processingattempt  (cost=8.02..13.36 rows=2 width=143)
  Recheck Cond: (case_ids @> '{14d34a63-b565-4925-a17c-416c901fdf9f}'::uuid[])
  ->  Bitmap Index Scan on generic_inb_case_id_434f37_gin (cost=0.00..8.02 rows=2 width=0)
        Index Cond: (case_ids @> '{14d34a63-b565-4925-a17c-416c901fdf9f}'::uuid[])

esoergel commented on corehq/motech/generic_inbound/models.py:
```diff
@@ -72,3 +76,50 @@ def to_json(self):
+    response_body = models.TextField()
+    raw_response = models.JSONField(default=dict)
+
+    xform_id = models.UUIDField(db_index=True, null=True, blank=True)
+    case_ids = ArrayField(models.UUIDField(default=uuid4), null=True, blank=True)
```
I'm thinking about switching these from `UUIDField` to `CharField`.  I like the idea of using UUIDs, and Django does normalize the form, which is pretty neat, but I think it's perhaps too far outside the norm of how we treat IDs in HQ.

esoergel replied:
@snopoke do you mean that you agree I should switch this back to `CharField`?

snopoke replied:
I think so, yes

esoergel commented on corehq/motech/generic_inbound/models.py:
```diff
@@ -72,3 +76,50 @@ def to_json(self):
+    class RequestMethod(models.TextChoices):
+        POST = 'POST'
+        PUT = 'PUT'
+        PATCH = 'PATCH'
```
I considered moving this to `motect/consts.py`, but ended up reasoning that a use-case-specific set of choices might make more sense.

esoergel commented on corehq/motech/generic_inbound/views.py:
```diff
@@ -157,7 +157,7 @@ def post(self, request, domain, **kwargs):
 @api_auth
 @api_throttle
-@require_http_methods(["POST"])
+@require_http_methods(list(RequestLog.RequestMethod))
```
note the commit message comment - this change makes the view allow POST, PUT, and PATCH requests.  I don't think we intend to treat those differently any time soon, and this seemed a reasonable change.

esoergel commented on corehq/motech/generic_inbound/views.py:
```diff
@@ -215,3 +225,37 @@ def _execute_case_api(domain, couch_user, device_id, context, api_model):
+    response_json = json.loads(response.content)
+    if is_success:
+        case_ids = [c['case_id'] for c in
+                    response_json.get('cases', [response_json.get('case')])]
```
this is what I'm talking about in my PR message - I don't love parsing this sort of info out from a response object, but will defer perfecting that until the "retry" PR.

snopoke replied:
A simpler option would be to make `_generic_inbound_api` return custom response object that includes the data you need

esoergel committed:
Tag models as cascading deletes

esoergel commented on corehq/motech/generic_inbound/models.py:
```diff
@@ -72,3 +83,57 @@ def to_json(self):
+        PATCH = 'PATCH'
+
+    id = models.UUIDField(primary_key=True, default=uuid4)
+    domain = models.CharField(max_length=255, db_index=True)
```
I wasn't aware there was a distinction between the two approaches, thank you.

esoergel replied:
Do you happen to know of any good resources on the distinction?  Like, are there circumstances where setting `db_index=True` is preferable?  From what I garnered in the stackoverflow post, it looks like the pattern ops index is included for inexact matches, but the interface doesn't make that clear at all, and I haven't been able to find any official Django docs on the difference.

snopoke replied:
Generally there shouldn't be a difference and the Meta class mechanism is used for adding more advanced indexes (like with multiple columns / conditions). But when you are using the PostgreSQL DB backend there is a special case: https://docs.djangoproject.com/en/4.1/ref/databases/#indexes-for-varchar-and-text-columns

esoergel commented on corehq/motech/generic_inbound/models.py:
```diff
@@ -72,3 +83,57 @@ def to_json(self):
+    case_ids = ArrayField(models.UUIDField(default=uuid4), null=True, blank=True)
+
+    class Meta:
+        indexes = [GinIndex(fields=['case_ids'])]
```
Think I should just drop it?  It would of course be nice to filter the report by case ID ("which requests updated this case"), but you could also get there by form ID, and that's the more fundamental transaction unit.  Like, if you're auditing a specific case, you'd find the form that made the change you're interested in, then come find the API request that submitted that form.

snopoke replied:
Yea, I think we can drop the index. But we should probably make sure that you can easily identify the API config from the form. Currently I don't think we're doing much in that regard.

esoergel replied:
Good point - I think a sensible way to approach that would be to use the `device_id` field for that.  That way, the XMLNS still identifies this as a case API submission, and the device ID would further clarify that this came from a particular configurable API.  We'd still have the user agent available in the log for auditing.  That seems in-line with the latest guidance on `submit_case_blocks`.

snopoke replied:
On nice addition would be the form name which is what shows up on in the UI if it's present.

esoergel commented on corehq/motech/generic_inbound/views.py:
```diff
@@ -214,3 +225,38 @@ def _execute_case_api(domain, couch_user, device_id, context, api_model):
+        error_message=response_body if not is_success else '',
+        username=request.couch_user.username,
+        request_method=request.method,
+        request_query=request.get_raw_uri(),
```
Looking at this again, I wonder if it should use `request.META.get("QUERY_STRING")`, since I don't think there's anything useful before the `?`

snopoke replied:
Yup, agree

esoergel committed:
Only store whitelisted headers

esoergel committed:
Create charfield indexes in Meta.indexes

https://github.com/dimagi/commcare-hq/pull/32204/files#r980993704

esoergel committed:
Switch UUIDField to CharField for familiarity

xforms and case IDs are typically treated as strings, not wrapped UUID
objects, so this should make them a little easier to work with

esoergel committed:
Remove GIN index

esoergel committed:
Store only querystring portion of request URI

snopoke commented on corehq/motech/generic_inbound/models.py:
```diff
@@ -72,3 +83,61 @@ def to_json(self):
+    request_query = models.CharField(max_length=8192)
+    request_body = models.TextField()
+    request_headers = models.JSONField(default=dict)
+    request_user_agent = models.TextField()
```
why is this separate from the headers?

esoergel replied:
I made it a top level field since it'll be needed for reprocessing, but I suppose that can also be extracted from headers at that point, and it probably won't be filtered on.  I'll remove it.

esoergel replied:
Oh actually it looks like it's not one of the auditcare standard request headers - auditcare normalizes user agent:
https://github.com/dimagi/commcare-hq/blob/d691d5b87d09aea74ef99c67065c606d3c0e8198/corehq/apps/auditcare/models.py#L73-L75
Sounds like options are:

* Keep it separate, as done here
* Follow auditcare's lead and split it out
* Add user agent to the list of "standard" headers and specifically exclude it from the auditcare use-case.

None of those sounds like an obviously awesome solution, but I think I favor #3, just pushed a go at that, but feel free to second-guess.

snopoke replied:
Option 3 seems fine to me. Thanks

esoergel committed:
put user agent with the other headers

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/configurable-api-logging
